src/driver/amdxdna/ve2_hwctx.c

```diff
@@ -89,6 +240,332 @@ static int ve2_create_host_queue(struct amdxdna_dev *xdna, struct ve2_hsa_queue
+
+	hsa_queue_pkt_set_valid(pkt);
+
+	return 0;
+}
+
+static int ve2_submit_cmd_single(struct amdxdna_ctx *hwctx, struct amdxdna_sched_job *job, u64 *seq)
+{
+	struct amdxdna_dev *xdna = hwctx->client->xdna;
+	struct amdxdna_gem_obj *cmd_bo = job->cmd_bo;
+	u32 cmd_data_len;
+	void *cmd_data;
+	int ret;
+
+	cmd_data = amdxdna_cmd_get_payload(cmd_bo, &cmd_data_len);
+	if (!cmd_data) {
+		XDNA_ERR(xdna, "Invalid command received in single cmd submit");
+		return -EINVAL;
+	}
+
+	if (get_ve2_dpu_data_next(cmd_data))
+		ret = submit_command_indirect(hwctx, cmd_data, seq);
+	else
+		ret = submit_command(hwctx, cmd_data, seq);
+	if (ret) {
+		XDNA_ERR(xdna, "Submit single command failed, error %d", ret);
+		return ret;
+	}
+
+	return ve2_hwctx_add_job(hwctx, job, *seq, 1);
+}
+
+static int ve2_submit_cmd_chain(struct amdxdna_ctx *hwctx, struct amdxdna_sched_job *job, u64 *seq)
+{
+	struct amdxdna_dev *xdna = hwctx->client->xdna;
+	struct amdxdna_gem_obj *cmd_bo = job->cmd_bo;
+	struct amdxdna_cmd_chain *cmd_chain;
+	u32 cmd_chain_len;
+	int ret;
+
+	cmd_chain = amdxdna_cmd_get_payload(cmd_bo, &cmd_chain_len);
+	if (!cmd_chain || cmd_chain_len < struct_size(cmd_chain, data, cmd_chain->command_count)) {
+		XDNA_ERR(xdna, "Invalid command received in cmd chain submit");
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < cmd_chain->command_count; i++) {
+		u32 boh = (u32)(cmd_chain->data[i]);
+		struct amdxdna_gem_obj *abo;
+		void *cmd_data;
+		u32 cmd_data_len;
+
+		abo = amdxdna_gem_get_obj(hwctx->client, boh, AMDXDNA_BO_CMD);
+		if (!abo) {
+			XDNA_ERR(xdna, "Failed to find cmd BO %d", boh);
+			return -ENOENT;
+		}
+
+		cmd_data = amdxdna_cmd_get_payload(abo, &cmd_data_len);
+		if (!cmd_data) {
+			XDNA_ERR(xdna, "Invalid command data received");
+			amdxdna_gem_put_obj(abo);
+			return -EINVAL;
+		}
+
+		if (get_ve2_dpu_data_next(cmd_data))
+			ret = submit_command_indirect(hwctx, cmd_data, seq);
+		else
+			ret = submit_command(hwctx, cmd_data, seq);
+		if (ret) {
+			XDNA_ERR(xdna, "Submit chain command(%d) failed, error %d", i, ret);
+			amdxdna_gem_put_obj(abo);
+			return ret;
+		}
+
+		amdxdna_gem_put_obj(abo);
+	}
+
+	return ve2_hwctx_add_job(hwctx, job, *seq, cmd_chain->command_count);
+}
+
+int ve2_cmd_submit(struct amdxdna_ctx *hwctx, struct amdxdna_sched_job *job, u32 *syncobj_hdls,
+		   u64 *syncobj_points, u32 syncobj_cnt, u64 *seq)
+{
+	struct amdxdna_dev *xdna = hwctx->client->xdna;
+	struct amdxdna_gem_obj *cmd_bo = job->cmd_bo;
+	int ret;
+	u32 op;
+
+	op = amdxdna_cmd_get_op(cmd_bo);
+	if (op != ERT_START_DPU && op != ERT_CMD_CHAIN) {
+		XDNA_WARN(xdna, "Unsupported ERT cmd: %d received", op);
+		return -EINVAL;
+	}
+
+	if (op == ERT_CMD_CHAIN)
+		ret = ve2_submit_cmd_chain(hwctx, job, seq);
+	else
+		ret = ve2_submit_cmd_single(hwctx, job, seq);
+
+	if (ret) {
+		XDNA_ERR(xdna, "Failed to submit a command. ret %d\n", ret);
+		return ret;
+	}
+
+	ve2_ring_doorbell(hwctx);
```

Comment:
The ve2_ring_doorbell() returns error code. But it is ignored. Please put a comment here, it can be one of below,
1. A TODO comment, the error will be handled in future commit.
2. A statement that ignore ve2_ring_doorbell()'s error and return 0 to caller is the expected design.